### ISSUE TITLE
`vcat` changes type of `PooledArray` columns

### ISSUE THREAD

**Issue by MrHenning** [JuliaData/DataFrames.jl#3061]
First, thank you all for this absolutely beautiful Julia Package :slightly_smiling_face: 

I am not sure if this is intended, but using `vcat` on `PooledArray` dataframe columns changes their type:

```julia
typeof(
	vcat(
		DataFrame(col=PooledArray(rand(1:2, 10))), 
		DataFrame(col=PooledArray(rand(1:2, 10)))
	).col
)
>>> Vector{Int64} (alias for Array{Int64, 1})
```

whereas it works with `CategoricalArrays`:
```julia
typeof(
	vcat(
		DataFrame(col=categorical(rand(1:2, 10))), 
		DataFrame(col=categorical(rand(1:2, 10)))
	).col
)
>>> CategoricalVector{Int64, UInt32, Int64, CategoricalValue{Int64, UInt32}, Union{}} (alias for CategoricalArray{Int64, 1, UInt32, Int64, CategoricalValue{Int64, UInt32}, Union{}})
```

Using `vcat` on `Vector`s keeps the `PooledArray` type:
```julia
typeof(
	vcat(
		PooledArray(rand(1:2, 10)), 
		PooledArray(rand(1:2, 10))
	)
)
>>> PooledVector{Int64, UInt8, Vector{UInt8}} (alias for PooledArray{Int64, UInt8, 1, Array{UInt8, 1}})
```

**Comment by nalimilan** [comment]
Yeah it sounds better to use `vcat` if possible. But AFAICT it's not completely trivial to do as we also support various options for when a column is missing from one of the inputs. This is another case where something like https://github.com/JuliaLang/julia/pull/20815 could be useful.

**Comment by bkamins** [comment]
@nalimilan to make this change we would have to materialize missing columns before `vcat`ting them.

As with current code switching to default `vcat` would result in:
```
julia> vcat([1,2,3], Iterators.repeated(missing, 5))
4-element Vector{Any}:
 1
 2
 3
  Base.Iterators.Take{Base.Iterators.Repeated{Missing}}(Base.Iterators.Repeated{Missing}(missing), 5)
```

The reason is that `vcat` treats iterators (like `Iterators.Repeated`) as scalars not as iterables.
(we would need to instead materialize `Iterators.Repeated` into something that would match type of other columns that are already present - as you noted - this is tricky)

**Comment by bkamins** [comment]
> Looks like we should improve `vcat` for `PooledArrays`?

Yes (a rabbit hole as usual)

**Comment by bkamins** [comment]
I am moving it to for a decision in 1.5 release since the decision is not one-sided.

@MrHenning - do you see significant benefits of keeping `PooledArray` under `vcat`?

**Comment by MrHenning** [comment]
> @MrHenning - do you see significant benefits of keeping `PooledArray` under `vcat`?

I often have a many rows in my DataFrames, so `PooledArray`s are my obvious go-to.
(I think I chose `PooledArrays` over `CategoricalArrays` because they play nicer with `Arrow.jl`, but I might be mistaken...)
That being said, right now I just re-convert the columns to `PooledArray`s *after* the `vcat`, which only adds a few lines of code.

So the benefit would be to have `vcat(::DataFrame, ::DataFrame)` behave like `vcat(::Vector, ::Vector)` (in that it keeps the Type), and reducing the lines of code you'd need to write.
On the other hand, it does seem to involve quite a bit of work to achieve this, so reverting to `CategoricalArrays` on the user side is probably the easier solution :slightly_smiling_face: 

In any case, thanks for your efforts!!! I'll still be happy if things stay as they are right now :slightly_smiling_face:
